doc/man-pages/man1/compile_et.1
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
'''   \*(M", \*(S", \*(N" and \*(T" are the equivalent of
'''   \*(L" and \*(R", except that they are used on ".xx" lines,
'''   such as .IP and .SH, which do another additional levels of
'''   double-quote interpretation
.ds M" """
.ds S" """
.ds N" """""
.ds T" """""
.ds L' '
.ds R' '
.ds M' '
.ds S' '
.ds N' '
.ds T' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds M" ``
.ds S" ''
.ds N" ``
.ds T" ''
.ds L' `
.ds R' '
.ds M' `
.ds S' '
.ds N' `
.ds T' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH COMPILE_ET 1 "OpenAFS" "18/May/2009" "AFS Command Reference"
.UC
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
compile_et \- Produce error text tables for compilation
.SH "SYNOPSIS"
\fBcompile_et\fR [\fB\-debug\fR] <<\ [\fB\-language\fR\ <\fIlang\fR] >>>
    <<\ [\fB\-prefix\fR\ <\fIprefix\fR] >>> [\fB\-v\fR <\fIversion\fR>] <\fIerror_table\fR>
.SH "DESCRIPTION"
The \fBcompile_et\fR command builds the error text tables for compilation.
This includes both a header file that contains a set of mappings between
error names and values and a \fI.c\fR (or \fI.msf\fR) file that provides a text
table of descriptions.
.PP
The <\fIerror_table\fR> argument specifies which error table to generate.
The error table specification should exist in the current working
directory or in the directory specified with \fB\-prefix\fR and should be
named \fIerror_table.et\fR.
.SH "CAUTIONS"
This command is used internally within the build process for OpenAFS.
Most users will access this information via the \fItranslate_et(1)\fR manpage rather than
via \fBcompile_et\fR.
.PP
This command does not use the standard AFS command-line parsing package.
.SH "OPTIONS"
.Ip "\fB\-debug\fR" 4
Does nothing.  It neither adds debugging information to the output nor
provides additional information on its operation.
.Ip "\fB\-language\fR <\fIlang\fR>" 4
Specifies the type of output to generate.  Currently, only \s-1ANSI\s0 C and K&R
are supported values (via the \fBc\fR and \fBk&r-c\fR values, respectively).
The default is \s-1ANSI\s0 C.  There is some support for \*(C+ started, but that is
not yet supported.
.Ip "\fB\-prefix <\fIprefix\fR\fR" 4
Specifies the directory to search for the \fIerror_table.et\fR file.
.Ip "\fB\-v\fR <\fIversion\fR>" 4
Specified the type of output file: valid values are 1 (the default, for C
files) or 2, for \fB.msf\fR file generation.
.SH "EXAMPLES"
The following command generates the files \fIpterror.h\fR and \fIpterror.c\fR, 
suitable for use with C programs:
.PP
.Vb 1
\&   % compile_et -p path/to/src/ptserver pterror
.Ve
The following command generates K&R style files instead:
.PP
.Vb 1
\&   % compile_et -p path/to/src/ptserver -lang 'k&r-c' pterror
.Ve
.SH "SEE ALSO"
the \fItranslate_et(1)\fR manpage
.SH "COPYRIGHT"
Copyright 2009 Steven Jenkins <steven@endpoint.com>
.PP
This documentation is covered by the IBM Public License Version 1.0.  This
man page was written by Steven Jenkins for OpenAFS.

.rn }` ''
.IX Title "COMPILE_ET 1"
.IX Name "compile_et - Produce error text tables for compilation"

.IX Header "NAME"

.IX Header "SYNOPSIS"

.IX Header "DESCRIPTION"

.IX Header "CAUTIONS"

.IX Header "OPTIONS"

.IX Item "\fB\-debug\fR"

.IX Item "\fB\-language\fR <\fIlang\fR>"

.IX Item "\fB\-prefix <\fIprefix\fR\fR"

.IX Item "\fB\-v\fR <\fIversion\fR>"

.IX Header "EXAMPLES"

.IX Header "SEE ALSO"

.IX Header "COPYRIGHT"



NEWS
OpenAFS News -- history of user-visible changes.

* Changes incorporated in OpenAFS 1.4.0

** Addition of pthreading support in volserver.

** Threading optimization of RX libraries.

** Windows 95/98 client desupported.

** Digital UNIX/Tru64 desupported.

** Windows XP/2003 client includes persistent (across boots) cache support.

** Windows XP/2003 now supports a cache of up to 1.2gb.

** Windows XP/2003 has various improvements to the included GUI and command 
   line tools.

** Windows XP/2003 now supports AFS UNC paths.

** MacOS X (10.3) and Linux use system inodes instead of a private inode pool.

** Files larger than 2gb are supported in Unix clients, and in servers if the 
   --enable-largefile-fileserver switch is enabled at configure time.

** Obsolete and insecure auxiliary network daemons (rcp, rsh, ftpd et al)
   have been removed.

** Fileserver optimizations have been made to improve timeout and callback 
   breaking behavior to inaccessible clients or clients whose address has
   changed.

** Kerberos 5 support is integrated.

** -nosettime is now the default for afsd.  Use "-settime" to get the
   old behavior.

** Mountpoint directory information is now only faked for cross-cell
   mountpoints when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd, but might not be stable
   on all platforms.

* Changes incorporated in OpenAFS 1.2.9

** The kaserver now defaults to not allowing interrealm authentication,
   due to security vulnerabilities in the krb4 protocol.  The new
   "-crossrealm" flag to the kaserver is provided to reenable interrealm
   authentication if desired.

** RedHat Linux 9.0 is now supported.

** Solaris 9 12/02 is now supported.  Solaris 7 and 8 x86 should now
   work again.

** On Linux machines using 2.2 series kernels, 2.2.19 or higher is now
   required.

** An OpenAFS 1.2.9 afsd will not work with kernel modules built from
   an earlier OpenAFS release.  In general, using a mismatched afsd and
   kernel modules set is unsupported; it is not recommended that you use
   such a configuration on a regular basis.

* Changes incorporated in OpenAFS 1.2.8

** Mountpoint directory information is now only faked for cross-cell
   mountpoits when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** HPUX 11.0 is now supported.

** It is now possible for AFS to use Kerberos 5 directly, via rxkad 2b.
   See the OpenAFS 1.2.8 Release Notes for more information on using this
   capability.

** An NFS translator kernel module is now included and compiled by default
   for Solaris only.

* Changes incorporated in OpenAFS 1.2.7

** MacOS X 10.2 is now supported.  FreeBSD 4.3 and later support is included
   in this release, but is still under active development and should only
   be used by those doing active development on the OpenAFS FreeBSD client.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** The fileserver will now use Rx pings to determine if clients are reachable
   prior to allocating resources to them, to prevent asymmetric clients from
   consuming all fileserver resources.

* Changes incorporated in OpenAFS 1.2.6

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd.

** Solaris 9 FCS and Solaris 7 and 8 x86 are now supported.

* Changes incorporated in OpenAFS 1.2.5

** A remote denial of service attack in the AIX and IRIX clients has
   been fixed.  Users of those platforms are strongly encouraged to
   upgrade.

** Fixed race conditions in fileserver that could result in crash.

* Changes incorporated in OpenAFS 1.2.4

** Server logfiles now more consistant about format in which hosts are
   referred to.

** vfsck on Solaris will now allow force runs (using -y flag) even if old 
   inodes exist.

* Changes incorporated in OpenAFS 1.2.3

** Cell aliases for dynroot can be specified in the CellAlias file in
   /usr/vice/etc or /usr/local/etc/openafs, in format "realname alias",
   one per line.  They can also be managed at runtime with "fs newalias"
   and "fs listaliases".

* Changes incorporated in OpenAFS 1.2.2

** Solaris 9 and Linux PA-RISC are now supported

** fileserver will not erroneously delay legitimate errors for 3 seconds
   after 10 errors are returned (e.g. stat() on a directory you can't read)

** Rx MTU calculation now works for Irix, Solaris and Linux

** If afsd is started with the -dynroot flag, /afs will be locally
   generated from the CellServDB.  AFSDB cells will be mounted
   automatically upon access.

** The namei fileserver allows vice "partitions" to be directories instead
   of partitions and will attach and display accordingly. Creating the file
   "AlwaysAttach" in the /vicepX directory is used as the trigger to attach it.

** TSM support for butc no longer requires editing a Makefile, simply
   specify the --enable-tivoli-tsm configure option.

** Linux builds no longer require source changes every time the kernel

src/comerr/compile_et.c
/*
 *
 * Copyright 1986, 1987, 1988
 * by MIT Student Information Processing Board.
 *
 * For copyright info, see "mit-sipb-cr.h".
 *
 */

#undef MEMORYLEAK
#include <afsconfig.h>
#include <afs/param.h>
#include <afs/afsutil.h>

RCSID
    ("$Header$");

#include <stdio.h>
#include <stdlib.h>

#ifndef AFS_NT40_ENV
#include <sys/file.h>
#include <sys/param.h>
#endif

#include <errno.h>
#include <string.h>
#ifdef	AFS_AIX32_ENV
#include <signal.h>
#endif
#include "mit-sipb-cr.h"
#include "internal.h"
#include "compiler.h"

#ifndef lint
static const char copyright[] =
    "Copyright 1987,1988 by MIT Student Information Processing Board";
#endif

extern char *current_token;
extern int table_number, current;
char buffer[BUFSIZ];
char *table_name = (char *)NULL;
char *file_table_name;
FILE *hfile, *cfile, *msfile;
int version = 1;
int use_msf = 0;

/* lex stuff */
extern FILE *yyin;
extern FILE *yyout;
#ifdef AFS_AIX51_ENV
int yylineno = 0;
#else
extern int yylineno;
#endif

char *
xmalloc(unsigned int size)
{
    char *p = (char *)malloc(size);
    if (!p) {
	perror(whoami);
	exit(1);
    }
    return p;
}

static int
check_arg(char const *const *str_list, char const *arg)
{
    while (*str_list)
	if (!strcmp(arg, *str_list++))
	    return 1;
    return 0;
}

static const char *const debug_args[] = {
    "d",
    "debug",
    0,
};

static const char *const lang_args[] = {
    "lang",
    "language",
    0,
};

static const char *const language_names[] = {
    "C",
    "K&R C",
    "C++",
    0,
};

static const char *const c_src_prolog[] = {
    "#include <afs/param.h>\n",
    "#include <afs/error_table.h>\n",
    0,
};

static const char *const krc_src_prolog[] = {
    "#ifdef __STDC__\n",
    "#define NOARGS void\n",
    "#else\n",
    "#define NOARGS\n",
    "#define const\n",
    "#endif\n\n",
    "#include <afs/param.h>\n",
    "#include <afs/error_table.h>\n",
    0,
};

static const char warning[] =
    "/*\n * %s:\n * This file is automatically generated; please do not edit it.\n */\n";

static const char msf_warning[] =
    "$ \n$ %s:\n$ This file is automatically generated; please do not edit it.\n$ \n$set 1\n";

/* pathnames */
char c_file[MAXPATHLEN];	/* output file */
char h_file[MAXPATHLEN];	/* output */
char msf_file[MAXPATHLEN];
char et_file[MAXPATHLEN];	/* full path to input file */

static void
usage(void)
{
    fprintf(stderr,
	    "%s: usage: %s ERROR_TABLE [-debug] [-language LANG] [-h INCLUDE] [-p prefix] [-v version]\n",
	    whoami, whoami);
    exit(1);
}

static void
dup_err(char const *type, char const *one, char const *two)
{
    fprintf(stderr, "%s: multiple %s specified: `%s' and `%s'\n", whoami,
	    type, one, two);
    usage();
}

#include "AFS_component_version_number.c"

int
main(int argc, char **argv)
{
    char *p, *ename;
    char const *const *cpp;
    int got_language = 0;
    char *got_include = 0;
    char *got_prefix = ".";

#ifdef	AFS_AIX32_ENV
    /*
     * The following signal action for AIX is necessary so that in case of a 
     * crash (i.e. core is generated) we can include the user's data section 
     * in the core dump. Unfortunately, by default, only a partial core is
     * generated which, in many cases, isn't too useful.
     */
    struct sigaction nsa;

    sigemptyset(&nsa.sa_mask);
    nsa.sa_handler = SIG_DFL;
    nsa.sa_flags = SA_FULLDUMP;
    sigaction(SIGSEGV, &nsa, NULL);
#endif
    /* argument parsing */
    debug = 0;
    filename = 0;
    whoami = argv[0];
    p = strrchr(whoami, '/');
    if (p)
	whoami = p + 1;
    while (argv++, --argc) {
	char *arg = *argv;
	if (arg[0] != '-') {
	    if (filename)
		dup_err("filenames", filename, arg);
	    filename = arg;
	} else {
	    arg++;
	    if (check_arg(debug_args, arg))
		debug++;
	    else if (check_arg(lang_args, arg)) {
		got_language++;
		arg = *++argv, argc--;
		if (!arg)
		    usage();
		if (language)
		    dup_err("languanges", language_names[(int)language], arg);

#define check_lang(x,v) if (!strcasecmp(arg,x)) language = v
		check_lang("c", lang_C);
		check_lang("ansi_c", lang_C);
		check_lang("ansi-c", lang_C);
		check_lang("krc", lang_KRC);
		check_lang("kr_c", lang_KRC);
		check_lang("kr-c", lang_KRC);
		check_lang("k&r-c", lang_KRC);
		check_lang("k&r_c", lang_KRC);
		check_lang("c++", lang_CPP);
		check_lang("cplusplus", lang_CPP);
		check_lang("c-plus-plus", lang_CPP);
#undef check_lang

		if (!language) {
		    fprintf(stderr, "%s: unknown language name `%s'\n",
			    whoami, arg);
		    fprintf(stderr, "\tpick one of: C K&R-C\n");
		    exit(1);
		}
	    } else if (strcmp(arg, "h") == 0) {
		arg = *++argv, argc--;
		if (!arg)
		    usage();
		got_include = arg;
	    } else if (strcmp(arg, "p") == 0) {
		arg = *++argv, argc--;
		if (!arg)
		    usage();
		got_prefix = arg;
	    } else if (strcmp(arg, "v") == 0) {
		arg = *++argv, argc--;
		version = atoi(arg);
		if (version != 1 && version != 2) {
		    fprintf(stderr, "%s: unknown control argument -`%s'\n",
			    whoami, arg);
		    usage();
		    exit(1);
		}
		if (version == 2)
		    use_msf = 1;
	    } else {
		fprintf(stderr, "%s: unknown control argument -`%s'\n",
			whoami, arg);
		usage();
	    }
	}
    }
    if (!filename)
	usage();
    if (!got_language)
	language = lang_C;
    else if (language == lang_CPP) {
	fprintf(stderr, "%s: Sorry, C++ support is not yet finished.\n",
		whoami);
	exit(1);
    }


    p = strrchr(filename, '/');
    if (p == (char *)NULL)
	p = filename;
    else
	p++;
    ename = xmalloc(strlen(p) + 5);
    strcpy(ename, p);

    /* Now, flush .et suffix if it exists.   */
    p = strrchr(ename, '.');
    if (p != NULL) {
	if (strcmp(p, ".et") == 0)
	    *p = 0;
    }

    if (use_msf) {
	sprintf(msf_file, "%s.msf", ename);
    } else {
	sprintf(c_file, "%s.c", ename);
    }
    if (got_include) {
	sprintf(h_file, "%s.h", got_include);
    } else {
	sprintf(h_file, "%s.h", ename);
    }
    p = strrchr(filename, '.');
    if (p == NULL) {
	p = xmalloc(strlen(filename) + 4);
	sprintf(p, "%s.et", filename);
	filename = p;
    }

    sprintf(et_file, "%s/%s", got_prefix, filename);

    yyin = fopen(et_file, "r");
    if (!yyin) {
	perror(et_file);
	exit(1);
    }

    /* on NT, yyout is not initialized to stdout */
    if (!yyout) {
	yyout = stdout;
    }

    hfile = fopen(h_file, "w");
    if (hfile == (FILE *) NULL) {
	perror(h_file);
	exit(1);
    }
    fprintf(hfile, warning, h_file);
    if (got_include) {
	char buffer[BUFSIZ];
	char prolog_h_file[MAXPATHLEN];
	FILE *prolog_hfile;
	int count, written;

	strcpy(prolog_h_file, got_prefix);
	strcat(prolog_h_file, "/");
	strcat(prolog_h_file, got_include);
	strcat(prolog_h_file, ".p.h");
	prolog_hfile = fopen(prolog_h_file, "r");
	if (prolog_hfile) {
	    fprintf(stderr, "Including %s at beginning of %s file.\n",
		    prolog_h_file, h_file);
	    fprintf(hfile, "/* Including %s at beginning of %s file. */\n\n",
		    prolog_h_file, h_file);
	    do {
		count =
		    fread(buffer, sizeof(char), sizeof(buffer), prolog_hfile);
		if (count == EOF) {
		    perror(prolog_h_file);
		    exit(1);
		}
		written = fwrite(buffer, sizeof(char), count, hfile);
		if (count != written) {
		    perror(prolog_h_file);
		    exit(1);
		}
	    } while (count > 0);
	    fprintf(hfile, "\n/* End of prolog file %s. */\n\n",
		    prolog_h_file);
	}
    }

    if (use_msf) {
	msfile = fopen(msf_file, "w");
	if (msfile == (FILE *) NULL) {
	    perror(msf_file);
	    exit(1);
	}
	fprintf(msfile, msf_warning, msf_file);
    } else {
	cfile = fopen(c_file, "w");
	if (cfile == (FILE *) NULL) {
	    perror(c_file);
	    exit(1);
	}
	fprintf(cfile, warning, c_file);

	/* prologue */
	if (language == lang_C)
	    cpp = c_src_prolog;
	else if (language == lang_KRC)
	    cpp = krc_src_prolog;
	else
	    abort();
	while (*cpp)
	    fputs(*cpp++, cfile);
    }

    /* parse it */
    yyparse();

    put_final_postlog();

    fclose(yyin);		/* bye bye input file */
    fclose(hfile);		/* bye bye include file */
    if (use_msf)
	fclose(msfile);
    return 0;
}

char *cot_suffix(int n)
{
    static char buf[16];
    if (!n) return "";
    sprintf(buf, "%d", n);
    return buf;
}

put_ecs_postlog()
{
    char lcname[6];

    if (!use_msf) {	/* { */
	char *cot = cot_suffix(count_of_tables);
	fputs("    0\n};\n\n", cfile);
	fprintf(cfile,
		"static const struct error_table et%s = { text%s, %ldL, %d };\n\n",
		cot, cot,
		(long int)table_number,
		current);
	fprintf(cfile,
		"static struct et_list etlink%s = { 0, &et%s};\n\n",
		cot, cot);


	if (!count_of_tables)
	    fprintf(hfile, "extern void initialize_%s_error_table(void);\n",
		    table_name);
    } else {
	if (!count_of_tables)
	    fprintf(hfile, "#define initialize_%s_error_table(void)\n",
		    table_name);
    }

    fprintf(hfile, "#define ERROR_TABLE_BASE_%s (%ldL)\n", table_name,
	    (long int)table_number);
    /* compatibility... */
    fprintf(hfile, "\n/* for compatibility with older versions... */\n");
    if (!count_of_tables)
	fprintf(hfile, "#define init_%s_err_tbl initialize_%s_error_table\n",
		table_name, table_name);
    fprintf(hfile, "#define %s_err_base ERROR_TABLE_BASE_%s\n", table_name,
	    table_name);
    lcstring(lcname, table_name, sizeof(lcname));
    if (strcmp(table_name, lcname)) {
	fprintf(hfile, "\n/* for compatibility with other users... */\n");
	fprintf(hfile, "#define ERROR_TABLE_BASE_%s (%ldL)\n", lcname,
		(long int)table_number);
	if (!count_of_tables) {
	    fprintf(hfile, "#define init_%s_err_tbl initialize_%s_error_table\n",
		    lcname, table_name);
	    fprintf(hfile,
		    "#define initialize_%s_error_table initialize_%s_error_table\n",
		    lcname, table_name);
	}
	fprintf(hfile, "#define %s_err_base ERROR_TABLE_BASE_%s\n", lcname,
		lcname);
    }
    if (!count_of_tables)
	file_table_name = table_name;
    ++count_of_tables;
}

put_final_postlog()
{
    int i;

    if (use_msf) return;
    fprintf(cfile, "void initialize_%s_error_table(void) {\n",
	    file_table_name);
    for (i = 0; i < count_of_tables; ++i) {
	fprintf(cfile, "    afs_add_to_error_table(&etlink%s);\n", cot_suffix(i));
    }
    fputs("}\n", cfile);
    fclose(cfile);
}

void
yyerror(const char *s)
{
    fputs(s, stderr);
    fprintf(stderr, "\nLine number %d; last token was '%s'\n", yylineno,
	    current_token);
}
